src/SAML2/Assertion.php

```diff
@@ -601,15 +615,15 @@ private function parseAttributeValue(DOMNode $attribute, string $attributeName):
             $this->attributesValueTypes[$attributeName][] = $type;
 
             if ($hasNonTextChildElements) {
-                $this->attributes[$attributeName][] = $value->childNodes;
+                $this->attributes[$attributeName]->addAttributeValue(
+                    new AttributeValue($value->childNodes)
+                );
                 continue;
             }
             
-            if ($type === 'xs:integer') {
```

Comment:
We used to allow attribute values to be an integer and then the xs:type would be properly set.. I think this PR breaks this behaviour..  It makes me wonder if the type should be just a property on the AttributeValue-class instead of an array on the Assertion-class?

Reply:
Similar, I think the FriendlyName should be a property of the Attribute-class...

Reply:
So all of the xsi:integer checks along with casts to ints currently exist in master Assertion.php.  So all of that is basically the same...not sure how this breaks anything.  
However I do notice that your fork's version of Assertion.php is very different from what I pushed to my fork years ago and tried to get into this PR.  Refer to the 'parseAttributeValue' function:
https://github.com/simplesamlphp/saml2/blob/2cba3c078149ef6f26b3610643fbf559aa46d737/src/SAML2/Assertion.php
I'm not sure what happened here, but somehow my changes didn't actually get pulled into your fork?

Reply:
You mean those `<<<<<<< HEAD` lines? Yeah, that's what I've been working on today to get rid of.. They're all gone now?

Regarding `xsi:integer`, your original PR relied on creating AttributeValues with integers, but now that we're fully typehinted that doesn't work anymore.. Only strings are allowed, but then we need to set the type somewhere and that appears to be missing..

Maybe we should focus on the main question and the issue will be taken care of anyway;  should `xsi:sometype` be a part of the AttributeValue-class instead of keeping an array on the Assertion-class?

Reply:
I would say yes - absolutely.

Reply:
OK. new starting point here:  https://github.com/simplesamlphp/saml2/tree/feature/friendlynames

Reply:
I agree, the type of the attribute needs to be a property of the `AttributeValue` class, and its serialization routines must take care of that (e.g. calling `intval()`).